src/agg/process_markers_symbolizer.cpp

```diff
@@ -79,9 +81,130 @@ struct agg_markers_renderer_context : markers_renderer_context
+            if (cacheable)
+            {
+                // Calculate canvas offsets
+                double margin = 0.0;
+                if (attrs[0].stroke_flag || attrs[0].stroke_gradient.get_gradient_type() != NO_GRADIENT)
+                {
+                    margin = std::abs(attrs[0].stroke_width);
+                }
+                double x0 = std::floor(src->bounding_box().minx() - margin);
+                double y0 = std::floor(src->bounding_box().miny() - margin);
+
+                // Now calculate subpixel offset and sample index
+                double dx = marker_tr.tx - std::floor(marker_tr.tx);
+                double dy = marker_tr.ty - std::floor(marker_tr.ty);
+
+                double sample_x = std::floor(dx * sampling_rate);
+                double sample_y = std::floor(dy * sampling_rate);
+
+                int sample_idx = static_cast<int>(sample_y) * sampling_rate + static_cast<int>(sample_x);
+
+                std::tuple<svg_path_ptr, svg_path_attrib_data, int> key(src, *reinterpret_cast<svg_path_attrib_data const*>(&attrs[0]), sample_idx);
+
+                auto it = cached_images_.find(key);
+                if (it == cached_images_.end())
+                {
+                    // Calculate canvas size
+                    int width  = static_cast<int>(std::ceil(src->bounding_box().width()  + 2.0 * margin)) + 2;
+                    int height = static_cast<int>(std::ceil(src->bounding_box().height() + 2.0 * margin)) + 2;
+
+                    // Reset clip box
+                    ras_.clip_box(0, 0, width, height);
+
+                    // Build local transformation matrix by resetting translation coordinates
+                    agg::trans_affine marker_tr_copy(marker_tr);
+                    marker_tr_copy.tx = sample_x / sampling_rate - x0;
+                    marker_tr_copy.ty = sample_y / sampling_rate - y0;
+
+                    // Create fill image
+                    std::shared_ptr<image_rgba8> fill_img;
+                    if (attrs[0].fill_flag || attrs[0].fill_gradient.get_gradient_type() != NO_GRADIENT)
+                    {
+                        fill_img = std::make_shared<image_rgba8>(width, height, true);
+
+                        agg::rendering_buffer buf(fill_img->bytes(), width, height, 4 * width);
+                        pixfmt_type pixf(buf);
+                        renderer_base renb(pixf);
+
+                        auto attrs_copy = attrs;
+                        attrs_copy[0].stroke_flag = false;
+                        attrs_copy[0].stroke_gradient.set_gradient_type(NO_GRADIENT);
+
+                        SvgRenderer svg_renderer(path, attrs_copy);
+                        render_vector_marker(svg_renderer, ras_, renb, src->bounding_box(), marker_tr_copy, 1.0f, false);
+
+                        if (std::all_of(fill_img->begin(), fill_img->end(), [](uint32_t val) { return val == 0; }))
```

Comment:
This feels really inefficient but I don't know if it's possible to use the `painted` attribute.